File: CSharp/Chain.cs
namespace com.dnene.josephus
{
    public class Chain
    {
        private Person first = null;

        public Chain(int size)
        {
            Person last = null;
            Person current = null;
            for (int i = 0 ; i < size ; i++)
            {
                current = new Person(i);
                if (first == null) first = current;
                if (last != null)
                {
                    last.Next = current;
                    current.Prev = last;
                }
                last = current;
            }
            first.Prev = last;
            last.Next = first;
        }

        public Person Kill(int nth)
        {
            Person current = first;
            int shout = 1;
            while(current.Next != current)
            {
                shout = current.Shout(shout, nth);
                current = current.Next;
            }
            first = current;
            return current;
        }

        public Person First => first;


    }

    public class Person
    {
        int count;
        private Person prev = null;
        private Person next = null;

        public Person(int count)
        {
            this.count = count;
        }

        public int Shout(int shout, int deadif)
        {
            if (shout < deadif) return (shout + 1);
            this.Prev.Next = Next;
            this.Next.Prev = Prev;
            return 1;
        }

        public int Count => count;

        public Person Prev
        {
            get => prev;
            set => prev = value;
        }
        public Person Next
        {
            get => next;
            set => next = value;
        }
    }

}

File: Python/Chain.py
class Person(object):
    def __init__(self,count):
        self.count = count;
        self.prev = None


        self.next = None
    def shout(self,shout,deadif):
        if (shout < deadif): return (shout + 1)
        self.prev.next = self.next
        self.next.prev = self.prev
        return 1

class Chain(object):
    def __init__(self,size):
        self.first = None
        last = None
        for i in range(size):
            current = Person(i)
            if self.first == None : self.first = current
            if last != None :
                last.next = current
                current.prev = last
            last = current
        self.first.prev = last
        last.next = self.first
    def kill(self,nth):
        current = self.first
        shout = 1
        while current.next != current:
            shout = current.shout(shout,nth)
            current = current.next
        self.first = current
        return current

import time

ITER = 100000

chains = [object() for i in range(ITER)]
target = [object() for i in range(ITER)]

start = time.perf_counter()

for i in range(ITER):
    chain = Chain(40)
    chain.kill(3)
    chains[i] = chain

for i in range(ITER):
    target[i] = chains[i]

end = time.perf_counter()

result_string = r"Time per iteration = {0} seconds ".format((end - start) / ITER )

print(result_string)
